Napari episode (#5)
bear-rsg/image-processing: episodes/10-multidimensional-data.md

```python
viewer.layers.clear()
viewer.add_image(data=cells, name="cells")
```

![](fig/cells-napari.png){alt='cells napari ss'}

:::::::::::::::::::::::::::::::::::::::: callout

## Clearing the Napari Viewer

When we want to close all Layers in a Viewer instance we can use `viewer.layers.clear()`. This
allows us to start fresh, alternatively we could close the existing Viewer and open a new one with
`viewer = napari.Viewer()`. We will use the first approach throughout this Episode but if any point
you accidentally close the Viewer you can just open a new one.

::::::::::::::::::::::::::::::::::::::::::::::::::

We can now scroll through the channels within Napari using the slider just below the image. This
approach will work with an arbitrary number of channels, not just three. With multichannel data it
is common to visualise colour overlays of selected channels where each channel has a different
colour map. In Napari we can do this programmatically using the `channel_axis` variable:

```python
viewer.layers.clear()
viewer.add_image(data=cells, name=["membrane", "cytoplasm", "nucleus"], channel_axis=0)
```

![](fig/cells-seperate-napari.png){alt='cells seperate napari ss'}

> Using what we have learnt in the previous Episodes lets segment the nuclei . First we produce a

Reviewer: Pedantry alert; should that be let's?

> Using what we have learnt in the previous Episodes lets segment the nuclei . First we produce a

Author: ```suggestion
Using what we have learnt in the previous Episodes let's segment the nuclei . First we produce a
```

binary mask of the nuclei by blurring the nuclei channel image and thresholding with the Otsu

> approach. Subsequently, we label the connected components within the mask to indentify individual

Reviewer: ```suggestion
approach. Subsequently, we label the connected components within the mask to identify individual
```

```

![](fig/nuc-napari.png){alt='nuclei napari ss'}

We can also interactively annotate images with shapes using a `Shapes` Layer. Let's display all
three channels and then do this within the GUI:

```python
viewer.layers.clear()
viewer.add_image(data=cells, name="cells")
# the instructor will demonstrate how to add a `Shapes` Layer and draw around cells using polygons
```

![](fig/shapes-napari.png){alt='shapes napari ss'}

:::::::::::::::::::::::::::::::::::::::  challenge

## Using Napari as an image viewer (20 min)

In [the *Capstone challenge* episode](09-challenges.md) you made a function to count the number of
bacteria colonies in an image also producing a new image that highlights the colonies. Modify this
function to make use of Napari, as opposed to `Matplotlib`, to display both the original colony
image
and the segmented individual colonies. Test this new function on `"data/colonies-03.tif"`. If you
did not complete the *Capstone challenge* epsidode you
can start with function from the
[solution](09-challenges.md#colony-counting-with-minimum-size-and-automated-threshold-optional-not-included-in-timing).
Hints:

> 1. A `napari.Viewer` object should be parsed to the function as an input parameter.

Reviewer: ```suggestion
1. A `napari.Viewer` object should be passed to the function as an input parameter.
```

corresponding to a specific axial position. As an example lets work with
`skimage.data.cells3d()`which is a 3D fluorescence microscopy image stack. From
the [dataset documentation](https://scikit-image.org/docs/stable/api/skimage.data.html#skimage.data.cells3d)
we can see that the data has two channels (0: membrane, 1: nuclei). The dimensions are ordered
`(z, channels, y, x)` and each voxel has a size of `(0.29 0.26 0.26)` microns. A voxel is the 3D
equivalent of a pixel and the size specifies
the physical dimensions, in `(z, y, x)` order for this example. Note the size of the voxels in z (
axial) is larger than the voxel spacing in the xy dimensions (lateral). Let's load the data and
visualise with Napari:

```python
cells3d = ski.data.cells3d()
viewer.layers.clear()
viewer.add_image(data=cells3d, name=["membrane", "nucleus"], channel_axis=1)
print(cells3d.shape)
```

```output
(60, 2, 256, 256)
```

![](fig/cells3d-slice-napari.png){alt='cell3d slice napari ss'}

Note we now have a dimension slice to control which slice we are visualizing. You can also switch
to a 3D rendering of the data:

![](fig/cells3d-volume-napari.png){alt='cell3d volume napari ss'}

> Many of the scikit-image functions you have used throughout this Lesson work with 3D (or indeed nD)

Reviewer: Not sure about the capital L for Lesson? Entirely up to you

Author: Yes capital L for a Carpentries "Lesson" which is their word for the whole course

image data. For example lets blur the nuclei channel using a 3D Gaussian filter and add the result
as a `Image` Layer to the Viewer:

```python
# extract the nuclei channel
nuclei = cells3d[:, 1, :, :]
# store the voxel size as a NumPy array (in microns)
voxel_size = np.array([0.29, 0.26, 0.26])
# get sigma (std) values for each dimension that corresponds to 0.5 microns
sigma = 0.5 / voxel_size
# blur data with 3D Guassian filter
blurred_nuclei = ski.filters.gaussian(nuclei, sigma=sigma)
# add to Napari Viewer
viewer.add_image(data=blurred_nuclei, name="nucleus_blurred")
print(sigma)
```

```output
[1.72413793 1.92307692 1.92307692]
```

![](fig/cells3d-blurred-napari.png){alt='cell3d volume napari ss'}

> Note we have used a different sigma for each dimension which corresponds to 0.5 microns in real

Reviewer: ```suggestion
Note we have used a different sigma for the z dimension to allow for the different voxel dimensions, to produce a blur of 0.5 microns in each axis of real space.
```
